nomad/state/state_store_host_volumes.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package state
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// HostVolumeByID retrieve a specific host volume
+func (s *StateStore) HostVolumeByID(ws memdb.WatchSet, ns, id string, withAllocs bool) (*structs.HostVolume, error) {
+	txn := s.db.ReadTxn()
+	obj, err := txn.First(TableHostVolumes, indexID, ns, id)
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil {
+		return nil, nil
+	}
+	vol := obj.(*structs.HostVolume)
+	if !withAllocs {
+		return vol, nil
+	}
+
+	vol = vol.Copy()
+	vol.Allocations = []*structs.AllocListStub{}
+
+	// we can't use AllocsByNodeTerminal because we only want to filter out
+	// allocs that are client-terminal, not server-terminal
+	allocs, err := s.AllocsByNode(nil, vol.NodeID)
+	if err != nil {
+		return nil, fmt.Errorf("could not query allocs to check for host volume claims: %w", err)
+	}
+	for _, alloc := range allocs {
+		if alloc.ClientTerminalStatus() {
+			continue
+		}
+		for _, volClaim := range alloc.Job.LookupTaskGroup(alloc.TaskGroup).Volumes {
+			if volClaim.Type == structs.VolumeTypeHost && volClaim.Source == vol.Name {
+				vol.Allocations = append(vol.Allocations, alloc.Stub(nil))
+			}
+		}
+	}
+
+	return vol, nil
+}
+
+// UpsertHostVolumes upserts a set of host volumes
+func (s *StateStore) UpsertHostVolumes(index uint64, volumes []*structs.HostVolume) error {
+	txn := s.db.WriteTxn(index)
+	defer txn.Abort()
+
+	for _, v := range volumes {
+		if exists, err := s.namespaceExists(txn, v.Namespace); err != nil {
+			return err
+		} else if !exists {
+			return fmt.Errorf("host volume %s is in nonexistent namespace %s", v.ID, v.Namespace)
+		}
+
+		obj, err := txn.First(TableHostVolumes, indexID, v.Namespace, v.ID)
+		if err != nil {
+			return err
+		}
+		if obj != nil {
+			// Only some properties of a volume can be updated in place.
+			//
+			// TODO(1.10.0): we need to check this in the RPC handler before
+			// sending the ClientRPC, so figure out if we really need to do this
+			// here.
```

Comment:
Hmm this looks like something an RPC-level validation should catch before it ever reaches the state store call indeed. But I'm ok with an extra check.